File: code/stir-bayes-backend.R
################################################################################
# stir-bayes-backend.R
# Bayesian models linking STIR and water-quality concentrations using rethinking
# Kerbel Long-Term Tillage Impacts Project
# Created by AJ Brown
#
# This script:
#   1. Loads long-format water-quality data with attached STIR metrics.
#   2. Cleans and formats the data for modeling (per-analyte structure).
#   3. Specifies and fits a simple hierarchical ulam model that regresses
#      log(concentration) on standardized STIR with analyte-specific slopes
#      and intercepts.
################################################################################

# ---- 0. Setup ----

# Load required packages
library(rethinking)
library(dplyr)
library(readr)
library(tidyr)
library(ggplot2)
library(blastula)

# If you prefer not to rely on setwd, open the Rproj at repo root and run there.
# Then all paths below are relative to the project root.

# ---- 1. Import WQ + STIR data ----

#' Load merged WQ × STIR data
#'
#' @param path Path to wq_with_stir_by_season.csv relative to project root.
#' @param year_max Optional upper bound on Year (e.g., 2023). Use Inf to keep all.
#' @return A tibble with raw merged WQ × STIR data.
load_wq_stir <- function(path = "out/pipeline_csvs/wq_with_stir_by_season.csv",
                         year_max = Inf) {
  
  # Resolve path robustly relative to current working directory
  if (!file.exists(path)) {
    alt <- file.path("..", path)
    if (file.exists(alt)) {
      path <- alt
    } else {
      stop(
        "File '", path, "' not found in '", getwd(), 
        "' or its parent directory."
      )
    }
  }
  
  dat <- readr::read_csv(path, show_col_types = FALSE)
  
  if (is.finite(year_max)) {
    dat <- dat %>% dplyr::filter(Year < year_max)
  }
  
  dat
}

# ---- 2. Data Cleaning ----

#' Clean merged WQ × STIR data
#'
#' Handles WQ flags ("U", "NA", "NA.IRR"), standardizes types,
#' and enforces factor/Date/numeric structure for modeling.
#'
#' @param wq_stir Tibble returned by load_wq_stir()
#' @return Cleaned tibble ready for modeling.
clean_wq_stir <- function(wq_stir) {
  
  wq_stir %>%
    # STEP 2A. Handle WQ flags systematically
    #   "NA"      = sample existed but was not measured      → keep as NA
    #   "U"       = nondetect (below detection limit)        → set to 0 for now
    #   "NA.IRR"  = no runoff occurred                       → drop rows
    dplyr::mutate(
      Result_mg_L        = ifelse(Result_mg_L == "u", "U", Result_mg_L),
      Inflow_Result_mg_L = ifelse(Inflow_Result_mg_L == "u", "U", Inflow_Result_mg_L)
    ) %>%
    # remove no-runoff cases entirely, but keep pure NAs
    dplyr::filter(is.na(Result_mg_L) | Result_mg_L != "NA.IRR") %>%
    # then your case_when and type conversions as before
    dplyr::mutate(
      Result_mg_L = dplyr::case_when(
        Result_mg_L == "U"  ~ "0",
        Result_mg_L == "NA" ~ NA_character_,
        TRUE                ~ Result_mg_L
      ),
      Inflow_Result_mg_L = dplyr::case_when(
        Inflow_Result_mg_L == "U"  ~ "0",
        Inflow_Result_mg_L == "NA" ~ NA_character_,
        TRUE                       ~ Inflow_Result_mg_L
      )
    ) %>%
    
    # STEP 2B. Explicit column type enforcement
    dplyr::mutate(
      # ---- Core numeric concentration fields ----
      Result_mg_L        = as.numeric(Result_mg_L),
      Inflow_Result_mg_L = as.numeric(Inflow_Result_mg_L),
      
      # ---- Identifiers ----
      Treatment  = factor(toupper(Treatment), levels = c("CT", "MT", "ST")),
      Rep        = factor(Rep),
      Analyte    = factor(Analyte),
      Irrigation = factor(Irrigation),
      InflowOutflow = factor(InflowOutflow),
      
      # ---- Dates ----
      Date        = as.Date(Date),
      PlantDate   = as.Date(PlantDate),
      HarvestDate = as.Date(HarvestDate),
      
      # ---- Season metadata ----
      SeasonYear = as.integer(SeasonYear),
      Crop       = factor(Crop),
      
      # ---- STIR predictors ----
      Season_STIR_toDate = as.numeric(Season_STIR_toDate),
      CumAll_STIR_toDate = as.numeric(CumAll_STIR_toDate),
      
      # ---- Boolean fields stored as text in pipeline ----
      Has_Inflow = dplyr::case_when(
        Has_Inflow == "TRUE"  ~ TRUE,
        Has_Inflow == "FALSE" ~ FALSE,
        TRUE                  ~ NA
      ),
      
      NoRunoff = dplyr::case_when(
        NoRunoff == "TRUE"  ~ TRUE,
        NoRunoff == "FALSE" ~ FALSE,
        TRUE                ~ NA
      ),
      
      # ---- Characterize other level variables as categorical ----
      Flag          = factor(Flag),
      Inflow_Flag   = factor(Inflow_Flag),
      FlumeMethod   = factor(FlumeMethod),
      MeasureMethod = factor(MeasureMethod),
      IrrMethod     = factor(IrrMethod),
      TSSMethod     = factor(TSSMethod),
      Lab           = factor(Lab),
      SampleMethod  = factor(SampleMethod),

      # ---- Flow/Volume ----
      Volume = as.numeric(Volume)
    ) %>%
    
    # STEP 2C. Create analyte_abbr for graphing + modeling
    dplyr::mutate(
      analyte_abbr = dplyr::case_when(
        Analyte == "Ammonium(NH4)"  ~ "NH4",
        Analyte == "ICP"            ~ "ICP",
        Analyte == "Nitrate"        ~ "NO3",
        Analyte == "NitrateNitrite" ~ "NOx",
        Analyte == "Nitrite"        ~ "NO2",
        Analyte == "NPOC"           ~ "NPOC",
        Analyte == "OrthoP"         ~ "OP",
        Analyte == "Selenium"       ~ "Se",
        Analyte == "TDS"            ~ "TDS",
        Analyte == "TKN"            ~ "TKN",
        Analyte == "TotalN"         ~ "TN",
        Analyte == "TotalP"         ~ "TP",
        Analyte == "TSP"            ~ "TSP",
        Analyte == "TSS"            ~ "TSS",
        TRUE                        ~ NA_character_
      ),
      analyte_abbr = factor(analyte_abbr,
                            levels = c("NH4","ICP","NO3","NOx","NO2","NPOC",
                                       "OP","Se","TDS","TKN","TN","TP","TSP",
                                       "TSS")
      )
    ) %>%
    
    # STEP 2D. Standardize OUT/INFLOW per-analyte concentrations
    dplyr::group_by(Analyte) %>%
    dplyr::mutate(
      cout_z = rethinking::standardize(Result_mg_L),
      cin_z  = rethinking::standardize(Inflow_Result_mg_L)
    ) %>%
    dplyr::ungroup() %>%
    
    # STEP 2E. Standardize STIR and volume metrics globally
    dplyr::mutate(
      stir_season_z = rethinking::standardize(Season_STIR_toDate),
      stir_cumall_z = rethinking::standardize(CumAll_STIR_toDate),
      volume_z      = rethinking::standardize(Volume)
    )
  
  
}

# ---- Use the functions ----

if (interactive()) {
  # Only run this section in interactive sessions
  # Step 1: load raw merged data
  wq_stir <- load_wq_stir(
    path = "out/pipeline_csvs/wq_with_stir_by_season.csv",
    year_max = Inf
  )
  
  # Quick peek at raw
  glimpse(wq_stir)
  
  # Step 2: clean and type-enforce
  wq_stir_clean <- clean_wq_stir(wq_stir)
}



# ---- Helper functions ----

# Create distance matrix for year effects
# Usage:
#   make_year_dist_mat(2011, 2025)              # contiguous years (old behavior)
#   make_year_dist_mat(unique(d_mod$Year))      # only observed years (recommended)
make_year_dist_mat <- function(start_or_years, end_year = NULL) {
  if (is.null(end_year)) {
    # Case 1: user passed a vector of years
    years <- sort(unique(start_or_years))
  } else {
    # Case 2: user passed start_year, end_year
    years <- seq(start_or_years, end_year)
  }
  
  n_years <- length(years)
  
  # pairwise absolute differences using outer
  dist_matrix <- abs(outer(years, years, "-"))
  
  rownames(dist_matrix) <- years
  colnames(dist_matrix) <- years
  
  dist_matrix
}


# cmdstan dashboard for convergence
cmdstan_dashboard <- function(
    fit,
    warmup = FALSE,
    plot = TRUE,
    trank = TRUE,
    vars = NULL,
    ess_col = c("ess_bulk","ess_tail")
) {
  
  if (!inherits(fit, "CmdStanMCMC")) {
    stop("cmdstan_dashboard(): 'fit' must be a CmdStanMCMC object from cmdstanr.")
  }
  ess_col <- match.arg(ess_col)
  
  # ---- sampler diagnostics (fast) ----
  x_raw <- fit$sampler_diagnostics(inc_warmup = warmup)
  
  # Force to plain base array (prevents S3 '[' dropping dims unexpectedly)
  x <- if (is.array(x_raw)) x_raw else tryCatch(unclass(x_raw), error = function(e) x_raw)
  if (!is.array(x) || length(dim(x)) < 2) {
    stop("cmdstan_dashboard(): sampler_diagnostics did not return an array-like object.")
  }
  
  # Expect dims: iter x chain x diag (but we guard)
  d <- dim(x)
  n_iter  <- d[1]
  n_chain <- d[2]
  n_samps <- n_iter * n_chain
  
  diag_names <- dimnames(x)[[3]]
  
  # Extract diag by name and force to n_iter x n_chain matrix
  diag_matrix <- function(name) {
    if (is.null(diag_names) || !(name %in% diag_names) || length(d) < 3) return(NULL)
    
    k <- match(name, diag_names)
    a <- x[, , k, drop = FALSE]   # iter x chain x 1 (ideally)
    a <- a[, , 1, drop = TRUE]    # should be iter x chain OR vector
    
    if (is.matrix(a) && all(dim(a) == c(n_iter, n_chain))) return(a)
    
    # If it collapsed to a vector, try to reshape if lengths match
    if (is.atomic(a) && length(a) == n_samps) {
      return(matrix(a, nrow = n_iter, ncol = n_chain))
    }
    
    NULL
  }
  
  div_mat   <- diag_matrix("divergent__")
  energy    <- diag_matrix("energy__")
  treedepth <- diag_matrix("treedepth__")
  n_leap    <- diag_matrix("n_leapfrog__")
  stepsize  <- diag_matrix("stepsize__")
  
  n_div <- if (!is.null(div_mat)) sum(div_mat) else NA_integer_
  
  # E-BFMI per chain (only if energy is truly matrix)
  ebfmi <- NULL
  if (is.matrix(energy)) {
    ebfmi <- sapply(seq_len(n_chain), function(ch) {
      e <- as.numeric(energy[, ch])
      num <- mean(diff(e)^2)
      den <- stats::var(e)
      if (is.finite(num) && is.finite(den) && den > 0) num / den else NA_real_
    })
    names(ebfmi) <- paste0("chain", seq_len(n_chain))
  }
  
  # ---- summary (can be slow if vars is NULL) ----
  s <- if (is.null(vars)) fit$summary() else fit$summary(variables = vars)
  
  rhat_vals <- if ("rhat" %in% names(s)) s$rhat else rep(NA_real_, nrow(s))
  ess_vals  <- if (ess_col %in% names(s)) s[[ess_col]] else rep(NA_real_, nrow(s))
  
  # ---- plotting ----
  if (plot) {
    oldpar <- par(no.readonly = TRUE)
    on.exit(par(oldpar), add = TRUE)
    par(mfrow = c(2,2), cex.axis = 0.9, cex.lab = 1)
    
    # Panel 1: ESS vs Rhat
    ok <- is.finite(rhat_vals) & is.finite(ess_vals) & ess_vals > 0
    y_rhat <- rhat_vals[ok]
    x_ess  <- ess_vals[ok]
    plot(
      x_ess, y_rhat,
      xlab = sprintf("%s effective sample size", ess_col),
      ylab = "Rhat",
      ylim = c(1, max(1.1, y_rhat, na.rm = TRUE)),
      log = "y",
      pch = 1,        # (1) hollow points
      cex = 0.6
    )
    abline(v = 0.1 * n_samps, col = "red")
    abline(v = n_samps, col = "gray60")
    abline(h = 1, lty = 2, col = "gray40")
    abline(h = 1.04, lty = 2, col = "gray40")  # (2) guide at 1.04
    
    # Panel 2: energy density (only if matrix)
    if (is.matrix(energy)) {
      e_all <- as.numeric(energy)
      d2 <- stats::density(
        e_all,
        adjust = 0.2,  # (3) less smoothing (was 0.8)
        na.rm = TRUE
      )
      plot(d2, main = "HMC energy", xlab = "energy__", ylab = "density")
      mu <- mean(e_all, na.rm = TRUE)
      sig <- stats::sd(e_all, na.rm = TRUE)
      if (is.finite(mu) && is.finite(sig) && sig > 0) {
        curve(stats::dnorm(x, mu, sig), add = TRUE, lwd = 2)
      }
      if (!is.null(ebfmi)) {
        mtext(
          sprintf("E-BFMI: %s", paste(sprintf("%.2f", ebfmi), collapse = ", ")),
          side = 3, line = 0.2, cex = 0.8
        )
      }
    } else {
      plot.new(); title("HMC energy")
      text(0.5, 0.5, "energy__ not available as iter×chain matrix", cex = 1.0)
    }
    
    # Panel 3: divergences
    plot.new(); title("Sampler issues")
    if (is.na(n_div)) {
      text(0.5, 0.75, "divergent__ not available", cex = 1.1)
    } else {
      text(0.5, 0.78, n_div, cex = 4)
      text(0.5, 0.58, "Divergent transitions", cex = 1.1)
    }
    
    # ---- Panel 4: lp__ trace plot (CmdStanR-safe, base R) ----
    trace_pars <- "lp__"
    
    if (!requireNamespace("posterior", quietly = TRUE)) {
      plot.new()
      title("Trace plot")
      text(0.5, 0.5, "Install the 'posterior' package to enable trace plots.", cex = 0.9)
    } else {
      
      dr <- fit$draws(
        variables = trace_pars,
        inc_warmup = warmup,
        format = "draws_array"
      )
      
      d_draws <- dim(dr)
      n_iter_draws  <- d_draws[1]
      n_chain_draws <- d_draws[2]
      var_names <- dimnames(dr)$variable
      
      v <- var_names[1]
      mat <- dr[, , v, drop = TRUE]
      if (!is.matrix(mat)) {
        mat <- matrix(mat, nrow = n_iter_draws, ncol = n_chain_draws)
      }
      
      chain_cols <- grDevices::hcl.colors(n_chain_draws, "Dark 3")
      ylim <- range(mat, finite = TRUE)
      
      plot(
        NA, xlim = c(1, n_iter_draws), ylim = ylim,
        xlab = "Iteration", ylab = v,
        main = paste0("Trace: ", v),
        bty = "l"
      )
      
      for (ch in seq_len(n_chain_draws)) {
        lines(seq_len(n_iter_draws), mat[, ch], col = chain_cols[ch], lwd = 0.8)
      }
      
      legend(
        "topright",
        legend = paste0("chain ", seq_len(n_chain_draws)),
        col = chain_cols,
        lty = 1,
        cex = 0.7,
        bty = "n"
      )
    }
  }
  
  invisible(list(
    sampler_diagnostics = x_raw,
    summary = s,
    n_samples = n_samps,
    n_divergent = n_div,
    ebfmi = ebfmi,
    stepsize = stepsize,
    n_leapfrog = n_leap
  ))
}

# email notification helper (requires mailR package and SMTP setup)
# Optional email notification after long sampling runs.
# To enable, define these environment variables before knitting/running:
#   STAN_NOTIFY_TO
#   STAN_NOTIFY_FROM
#   STAN_NOTIFY_USER
#   STAN_NOTIFY_PASS
#   STAN_NOTIFY_HOST   (for example: smtp.gmail.com)
#   STAN_NOTIFY_PORT   (for example: 465 or 587)
# This function is safe to leave in place when no SMTP credentials are set.

notify_sampling_done <- function(subject = NULL, body = NULL) {
  
  required <- c(
    "STAN_NOTIFY_TO",
    "STAN_NOTIFY_FROM",
    "STAN_NOTIFY_USER",
    "STAN_NOTIFY_PASS",
    "STAN_NOTIFY_HOST",
    "STAN_NOTIFY_PORT"
  )
  
  vals <- Sys.getenv(required, unset = "")
  names(vals) <- required
  
  if (any(vals == "")) {
    missing_vars <- names(vals)[vals == ""]
    message(
      "Sampling-finished email skipped: missing environment variable(s): ",
      paste(missing_vars, collapse = ", ")
    )
    return(invisible(FALSE))
  }
  
  if (!requireNamespace("blastula", quietly = TRUE)) {
    message("Sampling-finished email skipped: install.packages('blastula') is required.")
    return(invisible(FALSE))
  }
  
  smtp_port <- suppressWarnings(as.integer(Sys.getenv("STAN_NOTIFY_PORT")))
  if (is.na(smtp_port)) {
    message("Sampling-finished email skipped: STAN_NOTIFY_PORT is not a valid integer.")
    return(invisible(FALSE))
  }
  
  use_ssl <- identical(smtp_port, 465L)
  
  if (is.null(subject)) {
    subject <- sprintf("Stan sampling finished: %s", model_version)
  }
  
  if (is.null(body)) {
    body <- paste0(
      "Your CmdStan fit finished sampling.\n\n",
      "Model version: ", model_version, "\n",
      "Time finished: ", format(Sys.time(), "%Y-%m-%d %H:%M:%S %Z")
    )
  }
  
  email_obj <- blastula::compose_email(
    body = blastula::md(body)
  )
  
  tryCatch({
    blastula::smtp_send(
      email = email_obj,
      from = Sys.getenv("STAN_NOTIFY_FROM"),
      to = Sys.getenv("STAN_NOTIFY_TO"),
      subject = subject,
      credentials = blastula::creds_envvar(
        user = Sys.getenv("STAN_NOTIFY_USER"),
        pass_envvar = "STAN_NOTIFY_PASS",
        host = Sys.getenv("STAN_NOTIFY_HOST"),
        port = smtp_port,
        use_ssl = use_ssl
      )
    )
    
    message("Sampling-finished email sent.")
    invisible(TRUE)
    
  }, error = function(e) {
    message("Sampling-finished email failed: ", conditionMessage(e))
    invisible(FALSE)
  })
}

File: code/stir_bayes_backend.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
stir_bayes_backend.py

Python equivalent of `stir-bayes-backend.R::clean_wq_stir()`.

This script is intentionally *analysis-specific* and should be called as an
additional pipeline step AFTER `wq_with_stir_by_season.csv` is created.

Implements (as in the R code):
- Handle WQ flag tokens: "U", "NA", "NA.IRR" (and lowercase "u")
- Drop "NA.IRR" rows (no runoff), but preserve true missing values
- Enforce types (dates, numeric columns, key categorical columns)
- Create analyte_abbr (matching your mapping)
- Standardize:
  * cout_z, cin_z per analyte
  * stir_season_z, stir_cumall_z, volume_z globally
Notes:
- Z-scores use sample SD (ddof=1) to match R's sd() / rethinking::standardize().
"""

from __future__ import annotations

import argparse
from pathlib import Path
from typing import Optional

import numpy as np
import pandas as pd




def _clamp01(series: pd.Series, eps: float = 1e-6) -> pd.Series:
    """Clamp numeric values to (eps, 1-eps). Useful for Beta models."""
    vals = pd.to_numeric(series, errors="coerce").astype(float)
    return pd.Series(np.clip(vals, eps, 1.0 - eps), index=series.index)
ANALYTE_ABBR = {
    "Ammonium(NH4)": "NH4",
    "ICP": "ICP",
    "Nitrate": "NO3",
    "NitrateNitrite": "NOx",
    "Nitrite": "NO2",
    "NPOC": "NPOC",
    "OrthoP": "OP",
    "Selenium": "Se",
    "TDS": "TDS",
    "TKN": "TKN",
    "TotalN": "TN",
    "TotalP": "TP",
    "TSP": "TSP",
    "TSS": "TSS",
}
ANALYTE_ABBR_LEVELS = ["NH4","ICP","NO3","NOx","NO2","NPOC","OP","Se","TDS","TKN","TN","TP","TSP","TSS"]


def _zscore(x: pd.Series) -> pd.Series:
    """Sample z-score ignoring missing, like rethinking::standardize()."""
    xv = pd.to_numeric(x, errors="coerce").astype(float)
    mu = np.nanmean(xv)
    sd = np.nanstd(xv, ddof=1)
    if not np.isfinite(sd) or sd == 0:
        return (xv - mu) * np.nan
    return (xv - mu) / sd


def _token_series(s: pd.Series) -> pd.Series:
    """
    Convert series to cleaned token strings while preserving true missing.
    We want to keep literal tokens "U", "NA", "NA.IRR" for logic, and preserve NA as missing.
    """
    # If already numeric, convert to string but keep NaN as NaN
    if pd.api.types.is_numeric_dtype(s):
        return s.astype("float").astype("string")

    # Otherwise treat as string, but preserve missing
    ss = s.astype("string")
    ss = ss.str.strip()
    # normalize lowercase u -> U
    ss = ss.replace({"u": "U"})
    return ss


def _token_to_numeric(s: pd.Series) -> pd.Series:
    """Apply token rules: U -> 0, NA -> missing, otherwise numeric."""
    ss = _token_series(s)
    # Preserve true missing as missing
    # Replace tokens
    ss = ss.replace({"U": "0", "NA": pd.NA})
    return pd.to_numeric(ss, errors="coerce")


def _to_bool_best_effort(s: pd.Series) -> pd.Series:
    """
    Mirror R case_when that expects "TRUE"/"FALSE", but be more robust:
    accepts TRUE/FALSE in any case, 1/0, yes/no, and already-boolean.
    """
    if pd.api.types.is_bool_dtype(s):
        return s.astype("boolean")

    ss = s.astype("string").str.strip()
    up = ss.str.upper()

    mapping = {
        "TRUE": True, "FALSE": False,
        "T": True, "F": False,
        "1": True, "0": False,
        "YES": True, "NO": False,
        "Y": True, "N": False,
    }
    out = up.map(mapping)
    return out.astype("boolean")


def clean_wq_stir(df: pd.DataFrame) -> pd.DataFrame:
    """
    Cleaning + feature engineering for the Bayes STIR pipeline.

    Backwards compatible behavior:
      - Legacy numeric columns (Result_mg_L, Inflow_Result_mg_L) are preserved exactly as before,
        including any nondetect-to-zero mapping that _token_to_numeric() performs.
      - Existing z-score columns (cout_z, cin_z, stir_season_z, stir_cumall_z, volume_z) are preserved.

    New outputs (for future left-censoring / LOD modeling):
      - Result_is_nd (0/1) and Result_mg_L_cens (Float64 with NDs as NA)
      - Inflow_is_nd (0/1) and Inflow_Result_mg_L_cens (Float64 with NDs as NA)
      - Result_lod_mg_L and Inflow_lod_mg_L (RL preferred, else MDL)

    New outputs (to move standardization out of Stan):
      - IRR_z: global z-score of Irrigation
      - inflow_volume_z: global z-score of Inflow_Volume
    """
    out = df.copy()

    # --- STEP 2A: Handle WQ flags ("u","U","NA","NA.IRR") ---
    for col in ["Result_mg_L", "Inflow_Result_mg_L"]:
        if col in out.columns:
            out[col] = _token_series(out[col])

    # --- STEP 2A.1: Preserve ND information for censoring (backwards compatible) ---
    # We DO NOT modify legacy numeric columns here. We only create parallel *_cens columns and flags.
    def _nd_mask_from_tokens(s: pd.Series) -> pd.Series:
        # Match typical lab "U" qualifier. Adjust here if your token grammar changes.
        ss = s.astype("string").str.strip()
        return ss.str.upper().str.startswith("U")

    # Outflow ND flag + LOD
    if "Result_mg_L" in out.columns:
        tok = out["Result_mg_L"].astype("string").str.strip()
        nd_mask = _nd_mask_from_tokens(tok)
        out["Result_is_nd"] = nd_mask.fillna(False).astype(int)
        out["Result_mg_L_cens"] = pd.Series(pd.NA, index=out.index, dtype="Float64")

        # LOD: prefer RL, then MDL (both already in mg/L in the cleaned table)
        rl = pd.to_numeric(out["RL_mg_L"], errors="coerce") if "RL_mg_L" in out.columns else pd.Series(pd.NA, index=out.index)
        mdl = pd.to_numeric(out["MDL_mg_L"], errors="coerce") if "MDL_mg_L" in out.columns else pd.Series(pd.NA, index=out.index)
        out["Result_lod_mg_L"] = rl.combine_first(mdl).astype("Float64")

    # Inflow ND flag + LOD
    if "Inflow_Result_mg_L" in out.columns:
        tok = out["Inflow_Result_mg_L"].astype("string").str.strip()
        nd_mask = _nd_mask_from_tokens(tok)
        out["Inflow_is_nd"] = nd_mask.fillna(False).astype(int)
        out["Inflow_Result_mg_L_cens"] = pd.Series(pd.NA, index=out.index, dtype="Float64")

        rl = pd.to_numeric(out["RL_mg_L"], errors="coerce") if "RL_mg_L" in out.columns else pd.Series(pd.NA, index=out.index)
        mdl = pd.to_numeric(out["MDL_mg_L"], errors="coerce") if "MDL_mg_L" in out.columns else pd.Series(pd.NA, index=out.index)
        out["Inflow_lod_mg_L"] = rl.combine_first(mdl).astype("Float64")

    # Drop no-runoff cases entirely: Result_mg_L == "NA.IRR"
    if "Result_mg_L" in out.columns:
        rm = out["Result_mg_L"].astype("string").str.strip()
        # drop only explicit NA.IRR (case-insensitive, allowing stray spaces)
        drop_mask = rm.str.upper().eq("NA.IRR")
        out = out.loc[~drop_mask].copy()

    # Apply token -> numeric conversions (legacy behavior preserved)
    if "Result_mg_L" in out.columns:
        out["Result_mg_L"] = _token_to_numeric(out["Result_mg_L"])
        # Fill detected values into cens column (NDs remain NA)
        if "Result_mg_L_cens" in out.columns and "Result_is_nd" in out.columns:
            det = out["Result_is_nd"].fillna(0).astype(int).eq(0)
            out.loc[det, "Result_mg_L_cens"] = pd.to_numeric(out.loc[det, "Result_mg_L"], errors="coerce").astype("Float64")

    if "Inflow_Result_mg_L" in out.columns:
        out["Inflow_Result_mg_L"] = _token_to_numeric(out["Inflow_Result_mg_L"])
        if "Inflow_Result_mg_L_cens" in out.columns and "Inflow_is_nd" in out.columns:
            det = out["Inflow_is_nd"].fillna(0).astype(int).eq(0)
            out.loc[det, "Inflow_Result_mg_L_cens"] = pd.to_numeric(out.loc[det, "Inflow_Result_mg_L"], errors="coerce").astype("Float64")

    # --- STEP 2B: Type enforcement (best-effort, only if columns exist) ---
    if "Treatment" in out.columns:
        trt = out["Treatment"].astype("string").str.strip().str.upper()
        # Keep only expected levels; unknown -> NA
        trt = trt.where(trt.isin(["CT", "MT", "ST"]))
        out["Treatment"] = pd.Categorical(trt, categories=["CT", "MT", "ST"], ordered=True)

    for c in ["Rep", "Analyte", "Irrigation", "InflowOutflow", "Crop"]:
        if c in out.columns:
            out[c] = out[c].astype("string")

    for c in ["Date", "PlantDate", "HarvestDate"]:
        if c in out.columns:
            out[c] = pd.to_datetime(out[c], errors="coerce").dt.date

    if "SeasonYear" in out.columns:
        out["SeasonYear"] = pd.to_numeric(out["SeasonYear"], errors="coerce")

    # Numeric columns (include Irrigation because we compute IRR_z here)
    for c in ["Season_STIR_toDate", "CumAll_STIR_toDate", "Volume", "Inflow_Volume", "Irrigation"]:
        if c in out.columns:
            out[c] = pd.to_numeric(out[c], errors="coerce")

    for c in ["Has_Inflow", "NoRunoff"]:
        if c in out.columns:
            out[c] = _to_bool_best_effort(out[c])

    # --- STEP 2C: analyte_abbr ---
    if "Analyte" in out.columns:
        out["analyte_abbr"] = out["Analyte"].astype("string").map(ANALYTE_ABBR)
        out["analyte_abbr"] = pd.Categorical(out["analyte_abbr"], categories=ANALYTE_ABBR_LEVELS, ordered=True)

    # --- STEP 2D: Per-analyte standardization ---
    if "Analyte" in out.columns:
        if "Result_mg_L" in out.columns:
            out["cout_z"] = out.groupby("Analyte", dropna=False)["Result_mg_L"].transform(_zscore)
        if "Inflow_Result_mg_L" in out.columns:
            out["cin_z"] = out.groupby("Analyte", dropna=False)["Inflow_Result_mg_L"].transform(_zscore)

    # --- STEP 2E: Global standardization ---
    if "Season_STIR_toDate" in out.columns:
        out["stir_season_z"] = _zscore(out["Season_STIR_toDate"])
    if "CumAll_STIR_toDate" in out.columns:
        out["stir_cumall_z"] = _zscore(out["CumAll_STIR_toDate"])
    if "Volume" in out.columns:
        out["volume_z"] = _zscore(out["Volume"])

    # NEW: global irrigation z-score (replaces Stan transformed-data IRR_z)
    if "Irrigation" in out.columns:
        out["IRR_z"] = _zscore(out["Irrigation"])

    # NEW: global inflow volume z-score (candidate covariate for future model)
    if "Inflow_Volume" in out.columns:
        out["inflow_volume_z"] = _zscore(out["Inflow_Volume"])

    # --- Residue: percent cover to proportion for Beta modeling ---
    # Input expected on 0–100 scale as Residue_PercentCover.
    # We produce residue_prop in (0,1) (clamped away from 0/1 for Beta support).
    if "Residue_PercentCover" in out.columns:
        rp = pd.to_numeric(out["Residue_PercentCover"], errors="coerce") / 100.0
        out["residue_prop"] = _clamp01(rp)
        # 1 if residue was observed in the input table, else 0
        out["residue_obs"] = out["Residue_PercentCover"].notna().astype(int)

    return out
def main() -> None:
    ap = argparse.ArgumentParser(description="Bayes-specific cleaning for Kerbel STIR × WQ merged data.")
    ap.add_argument("--in", dest="src", default="out/pipeline_csvs/wq_with_stir_by_season.csv",
                    help="Input merged WQ×STIR CSV.")
    ap.add_argument("--out", dest="dst", default="out/wq_cleaned.csv",
                    help="Output cleaned CSV for Bayes modeling.")
    ap.add_argument("--debug", action="store_true")
    args = ap.parse_args()

    src = Path(args.src)
    dst = Path(args.dst)
    if not src.exists():
        raise FileNotFoundError(f"Input not found: {src}")

    # keep_default_na=False avoids pandas treating literal 'NA' as missing;
    # we handle tokens explicitly.
    df = pd.read_csv(src, keep_default_na=False)

    cleaned = clean_wq_stir(df)

    dst.parent.mkdir(parents=True, exist_ok=True)
    cleaned.to_csv(dst, index=False)

    print(f"[OK] Wrote Bayes-cleaned file → {dst}")
    if args.debug:
        print(f"[INFO] Rows in:  {len(df)}")
        print(f"[INFO] Rows out: {len(cleaned)}")
        for c in ["Has_Inflow", "NoRunoff"]:
            if c in cleaned.columns:
                vc = cleaned[c].value_counts(dropna=False)
                print(f"[INFO] {c} value_counts (incl NA):")
                print(vc.to_string())
        for c in ["cout_z", "stir_season_z", "volume_z"]:
            if c in cleaned.columns:
                x = cleaned[c].astype(float)
                print(f"[INFO] {c}: mean≈{np.nanmean(x):.4f} sd≈{np.nanstd(x, ddof=1):.4f}")


if __name__ == "__main__":
    main()
